pkg/cloudclient/gcp/private.go

```diff
@@ -0,0 +1,417 @@
+
+	c.logger.Debug(ctx, "Gathering description of instance type %s from EC2", c.instanceType)
+
+	descOut := c.computeService.MachineTypes.List(c.projectID, c.zone)
+
+	//check if instanceType is in the list
+	found := false
+	if err := descOut.Pages(ctx, func(page *computev1.MachineTypeList) error {
+		for _, machineType := range page.Items {
+			if string(machineType.Name) == c.instanceType {
+				found = true
+				c.logger.Debug(ctx, "Instance type %s supported", c.instanceType)
+				break
+			}
+		}
+		c.logger.Debug(ctx, "Fully describe instance types output contains %d instance types", len(page.Items))
+		return nil
+	}); err != nil {
+		return fmt.Errorf("Unable to gather list of supported instance types from ComputeService: %v", err)
+	}
+
+	if !found {
+		return fmt.Errorf("Instance type %s not found in ComputeService API", c.instanceType)
+	}
+
+	return nil
+}
+
+//ToDo func createComputeServiceInstance
+func (c *Client) createComputeServiceInstance(ctx context.Context, input createComputeServiceInstanceInput) (createComputeServiceInstanceInput, error) {
+
+	req := &computev1.Instance{
+		Name:        input.instanceName,
+		MachineType: fmt.Sprintf("zones/%s/machineTypes/%s", c.zone, input.machineType),
+
+		// Tags: &computev1.Tags{
+		// 	Items: []string{"http-server", "https-server"},
+		// },
+		Disks: []*computev1.AttachedDisk{
+			{
+				InitializeParams: &computev1.AttachedDiskInitializeParams{
+					DiskSizeGb:  10,
+					SourceImage: input.sourceImage,
+					// sourceImageEncryptionKey: &computepb.
+				},
+				AutoDelete: true,
+				Boot:       true,
+				Type:       "PERSISTENT",
+			},
+		},
+
+		NetworkInterfaces: []*computev1.NetworkInterface{
+			{
+				Name:       input.networkName,
+				Subnetwork: input.vpcSubnetID,
+			},
+		},
+		//pass gcpuserdata.yaml cloud-init script
+		Metadata: &computev1.Metadata{
+			Items: []*computev1.MetadataItems{
+				//can pass startup script
+				// {
+				// 	Key: proto.String("startup-script"),
+				// 	Value: proto.String("#!/bin/bash\n"),
+				// },
+
+				//pass gcpuserdata.yaml
+				{
+					Key:   "user-data",
+					Value: &input.userdata,
+				},
+			},
+		},
+	}
+
+	//send request to computeService
+	instanceResp, err := c.computeService.Instances.Insert(c.projectID, c.zone, req).Context(ctx).Do()
+	if err != nil {
+		return input, fmt.Errorf("unable to create instance: %v %v", err, instanceResp)
+	}
+
+	c.logger.Info(ctx, "Created instance with ID: %s", input.instanceName)
+
+	//get fingerprint from instance
+	inst, err := c.computeService.Instances.Get(c.projectID, c.zone, input.instanceName).Do()
+	if err != nil {
+		c.logger.Debug(ctx, "Failed to get fingerprint to apply tags to instance %v", err)
+	}
+
+	//Add tags - known as labels in gcp
+	c.logger.Info(ctx, "Applying labels")
+
+	reqbody := &computev1.InstancesSetLabelsRequest{
+		LabelFingerprint: inst.LabelFingerprint,
+		Labels:           c.tags,
+	}
+
+	//send request to apply tags, continue if tags are invalid
```

Comment:
are the tags really needed if we continue no matter what ?

Reply:
Fixed, now there's an error, we don't continue if tags are incorrect